client/src/main/java/uk/co/blackpepper/bowman/LinkedResourceMethodHandler.java

```diff
@@ -82,49 +82,64 @@ private Object invokeAnnotatedMethod(Object self, Method method, Method proceed,
 	}
 
 	private Object resolveLinkedResource(Object self, Method method, Method proceed, Object[] args)
-			throws IllegalAccessException, InvocationTargetException {
-		
+	throws IllegalAccessException, InvocationTargetException {
+
 		URI associationResource = new MethodLinkUriResolver(resource).resolveForMethod(method, args);
-		
+
 		if (Collection.class.isAssignableFrom(method.getReturnType())) {
 			Class<?> linkedEntityType = (Class<?>) ((ParameterizedType) method.getGenericReturnType())
-				.getActualTypeArguments()[0];
-			
-			return resolveCollectionLinkedResource(associationResource, linkedEntityType, self, proceed);
-		}
+					.getActualTypeArguments()[0];
 
+			if (proceed == null) {
+				return createCollectionOfLinkedResource(getLinkedResources(associationResource,
+						linkedEntityType), method);
+			}
+			else {
+				return createCollectionOfLinkedResource(getLinkedResources(associationResource, linkedEntityType),
+						self, proceed);
+			}
+		}
 		return resolveSingleLinkedResource(associationResource, method.getReturnType());
 	}
 
 	private <F> F resolveSingleLinkedResource(URI associationResource, Class<F> linkedEntityType) {
 		Resource<F> linkedResource = restOperations.getResource(associationResource, linkedEntityType);
-		
+
 		if (linkedResource == null) {
 			return null;
 		}
-		
+
 		return proxyFactory.create(linkedResource, restOperations);
 	}
 
-	private <F> Collection<F> resolveCollectionLinkedResource(URI associationResource, Class<F> linkedEntityType,
```

Comment:
My thinking with the refactoring was that without the Extract Methods the method was getting quite chunky and doing a fair few different things within it's body. And for the overloading of the method call the decision was made because I wanted the methods to actually use all the parameters that were passed to it, to avoid the need for two null checks in one method. I'm more than happy to change it to your preference but I do think the refactoring makes this section of code a bit more modular.

Reply:
OK, cool! I had been thinking the original design was more cohesive in having all the collection-resolving in one place, but see now that by creating the overloads you have reduced the need for optional params in those methods, which is also a good thing.

I think I would be happier if you renamed the two `createCollectionOfLinkedResource` overloads back to `resolveCollectionLinkedResource` as I think this provides a useful counterpoint to the `resolveLinkedResource`, `resolveSingleLinkedResource` method names. I think this would also make the method that immediately calls `updateCollectionWithLinkedResources` more clear (does creating something really update it?)

What do you reckon?

Reply:
yeah i agree, I was thinking the same thing as I was writing the code. Shall do that now and push the rest of those changes.